templates/fine-tune-function-calling/fc_utils/eval_utils.py

```diff
@@ -0,0 +1,163 @@
+    match = True
+    current_conv = []
+    assist_id = 0
+    conv_id = 0
+    while conv_id < len(messages):
+        if messages[conv_id]["role"] != "assistant":
+            current_conv.append(messages[conv_id])
+        # 'role' is 'assistant'. Thus, get assistant response
+        else:
+            if current_conv[-1]["role"] == "tool":
+                # If the last message was a tool, we need additional processing.
+                # For a tool, we need to have a valid tool call id. This can only be retrieved from the previous assistant response
+                # because our dataset is synthetically constructed
+                if (
+                    current_conv[-2]["role"] == "assistant"
+                    and current_conv[-2]["tool_calls"]
+                ):
+                    current_conv[-1]["tool_call_id"] = current_conv[-2]["tool_calls"][0].id
+                else:
+                    return current_conv, False
+            if isinstance(parser, OpenAIResponseParser):
+                parsed_response = parser.get_parsed_response(current_conv, tools)
+            else:
+                parsed_response = parser.get_parsed_response(current_conv)
+            if parsed_response["content"] == ERROR_OUTPUT:
+                return None, None  # return None if there's an error
+
+            is_correct, mistake_type = is_match(parsed_response, expected_responses[assist_id])
+            match = match and (is_correct)
+            original_assistant_response = parsed_response["original_response"]
+            current_conv.append(dict(original_assistant_response))
+            assist_id += 1 # next expected response
+            if not match:  # return right away if model output is incorrect
+                return current_conv, match, mistake_type
+        conv_id += 1 # next message
+    return current_conv, match, mistake_type
+
+
+def evaluate_gpt(ds: List[dict], openai_parser: OpenAIResponseParser):
+    """
+    Evaluates the GPT4 model on the test dataset.
+    """
+    num_correct = 0
+    total_count = 0
+    results = []
+    pbar = tqdm(total=len(ds), desc="Evaluating GPT4..")
+    for example in ds:
+        openai_messages = example["openai_messages"]
+        openai_tools = [
+            {"type": "function", "function": fn} for fn in example["tools"]
+        ]
+        openai_conv, openai_is_match, mistake_type = parse_and_eval(
+            openai_parser, openai_messages, example["expected_responses"], tools=openai_tools
+        )
+        pbar.update(1)
+        if openai_conv is None:  # skip if api errors out
+            openai_is_match = None
+        else:
+            total_count += 1
+        result_dict = {"correct": openai_is_match, "mistake_type": mistake_type, "example": example, "conv": openai_conv}
+        results.append(result_dict)
+
+        if openai_is_match:
+            num_correct += 1
+    print("GPT4 accuracy: ", num_correct / total_count)
```

Comment:
in jupyter notebooks, don't print stuff in functions and only print them in the notebooks